### DISCUSSION TITLE
[Question] How to perform better with Paratest?

### DISCUSSION THREAD

**Discussion by erusso7** [paratestphp/paratest#820] [Q&A]
Hello, I have a question:
Why is slower to user the paratest extension instead of phpunit ? 

I have 1049 unit tests and.... Look this 4 differents executions:

# Paratest:
```
$ vendor/bin/paratest -p10 -c phpunit.xml.dist
Running phpunit in 10 processes with <PROJECT_PATH>/vendor/bin/phpunit
Time: 12.54 seconds, Memory: 22.00MB
```
```
$ vendor/bin/paratest -c phpunit.xml.dist
Running phpunit in 5 processes with <PROJECT_PATH>/vendor/bin/phpunit
Time: 12.89 seconds, Memory: 22.00MB
```
```
$ vendor/bin/paratest -p1 -c phpunit.xml.dist
Running phpunit in 1 processes with <PROJECT_PATH>/vendor/bin/phpunit
Time: 26.68 seconds, Memory: 22.00MB
```

# PHPUnit
```
$ vendor/bin/phpunit -c phpunit.xml.dist
PHPUnit 6.5.6 by Sebastian Bergmann and contributors.
Time: 7.71 seconds, Memory: 70.00MB
```

**Comment by julianseeger** [comment]
You should take a look at the "optimizing speed" section of the readme.

About 1000 tests in 7s means 7ms per test. Paratest with the default config will spawn 1 process per Testcase which is 1000 process spawns in the worst case. 
Spawning a process is a pretty expensive task for the OS and may take far more than a ms, thus adding the 21s to your testrun with "-p1".

Paratest performs best for slow tests when the spawning overhead is compensated by the parallelization benefit (so -p1 will never give good results).
You may try using the WrapperRunner which spawns the processes once and then reuses them. When you combine that with a high number of processes (-p8 or more for a processor with 8 threads) and maybe -f to run even methods in parallel, that might give you a slight incease even with your fast tests.
